provider/vsphere/environ_policy.go

```diff
@@ -23,8 +24,51 @@ func (env *environ) PrecheckInstance(ctx context.ProviderCallContext, args envir
 
 // PrecheckInstance is part of the environs.Environ interface.
 func (env *sessionEnviron) PrecheckInstance(ctx context.ProviderCallContext, args environs.PrecheckInstanceParams) error {
-	_, err := env.parsePlacement(ctx, args.Placement)
-	return err
+	if _, err := env.parsePlacement(ctx, args.Placement); err != nil {
+		return errors.Trace(err)
+	}
+	if err := env.checkZones(ctx, args.Constraints.Zones); err != nil {
```

Comment:
And exit early if constraints not set?

Reply:
Constraints is a struct rather than a pointer, so it ends up being `args.Constraints.String() == ""` - do you think we need an early return here? Both of the checks return early if the zones/datastore are nil.